### PR TITLE
[kube-prometheus-stack] allow override of for and severity rules

### LINKED ISSUE #3384: [kube-prometheus-stack] Makes default prometheus rules more flexible
### Is your feature request related to a problem ?

By default, the charts create many useful prometheus rules. But for some, it could make sense to override params such `label.severity` or `for`.

Currently, those values are hardcoded so users have to disable the rule and "fork it" on their value.yml

### Describe the solution you'd like.

To allow "customization" on alerts, a map could be created and exposed on value.yml.

Example:
To change the `for` and `label.severity` from `KubePodNotReady` alert (https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/templates/prometheus/rules-1.14/kubernetes-apps.yaml#L62-L64)

on **values.yaml**
```
defaultRules:
  customize:
    KubePodNotReady:
      for: 5m
      labels:
        severity: critical
```

on **templates/prometheus/rules-1.14/kubernetes-apps.yaml**

```
    - alert: KubePodNotReady
...
      for: {{ .Values.defaultRules.customize.KubePodNotReady.for | default "15m" }}
      labels:
        severity: {{ .Values.defaultRules.customize.KubePodNotReady.labels.severity | default "warning" }}
```

Of course, the structure/naming needs to be thought to allow more advanced usecases (add more labels on the alert, change some parts of expr ...)

### Describe alternatives you've considered.

NONE

### Additional context.

issues linked:
* https://github.com/prometheus-community/helm-charts/issues/2409
* https://github.com/prometheus-community/helm-charts/issues/2662

### PR DESCRIPTION
**Pull request by ps-xaf**
#### What this PR does / why we need it

Attemp to address the issue with fixed values "for" and "severity" in the helm chart, to make it more flexible to use and allow to overload those values without forking and customzing the entire chart. 

#### Which issue this PR fixes

- fixes #3384 

#### Special notes for your reviewer
```
$ docker run -it --rm --network host --workdir=/data --volume ~/.kube/config:/root/.kube/config:ro --volume $(pwd):/data --volume ~/.config/helm:/root/.config/helm quay.io/helmpack/chart-testing:v3.10.1 ct lint --chart-dirs . --charts . --validate-maintainers=false                                                                                                                                   Linting charts...
Version increment checking disabled.

------------------------------------------------------------------------------------------------------------------------
 Charts to be processed:
------------------------------------------------------------------------------------------------------------------------
 kube-prometheus-stack => (version: "56.6.3", path: ".")
------------------------------------------------------------------------------------------------------------------------

Hang tight while we grab the latest from your chart repositories...
... 
</>snip</>
...
Update Complete. ⎈Happy Helming!⎈
Saving 5 charts
Dependency crds did not declare a repository. Assuming it exists in the charts directory
Downloading kube-state-metrics from repo https://prometheus-community.github.io/helm-charts
Downloading prometheus-node-exporter from repo https://prometheus-community.github.io/helm-charts
Downloading grafana from repo https://grafana.github.io/helm-charts
Downloading prometheus-windows-exporter from repo https://prometheus-community.github.io/helm-charts
Deleting outdated charts
Linting chart "kube-prometheus-stack => (version: \"56.6.3\", path: \".\")"
Validating /data/Chart.yaml...
Validation success! 👍

Linting chart with values file "ci/01-provision-crds-values.yaml"...

==> Linting .

1 chart(s) linted, 0 chart(s) failed

Linting chart with values file "ci/02-test-without-crds-values.yaml"...

==> Linting .

1 chart(s) linted, 0 chart(s) failed

Linting chart with values file "ci/03-non-defaults-values.yaml"...

==> Linting .

1 chart(s) linted, 0 chart(s) failed

Linting chart with values file "ci/04-prometheus-operator-webhook-values.yaml"...

==> Linting .

1 chart(s) linted, 0 chart(s) failed

Linting chart with values file "ci/05-alertmanager-ingress-host-with-wildcard-values.yaml"...

==> Linting .

1 chart(s) linted, 0 chart(s) failed

------------------------------------------------------------------------------------------------------------------------
 ✔︎ kube-prometheus-stack => (version: "56.6.3", path: ".")
------------------------------------------------------------------------------------------------------------------------
All charts linted successfully
```

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

File: charts/kube-prometheus-stack/templates/prometheus/rules-1.14/alertmanager.rules.yaml [hunk 1]
```diff
@@ -42,12 +42,12 @@ spec:
...
         # https://www.robustperception.io/alerting-on-gauges-in-prometheus-2-0 for details.
         max_over_time(alertmanager_config_last_reload_successful{job="{{ $alertmanagerJob }}",namespace="{{ $namespace }}"}[5m]) == 0
-      for: 10m
+      for: {{ dig "AlertmanagerFailedReload" "for" "10m" .Values.customRules }}
       {{- with .Values.defaultRules.keepFiringFor }}
       keep_firing_for: "{{ . }}"
       {{- end }}
       labels:
-        severity: critical
+        severity: {{ dig "AlertmanagerFailedReload" "labelsSeverity" "critical" .Values.customRules }}
```
Comment (jkroepke) [inline charts/kube-prometheus-stack/templates/prometheus/rules-1.14/alertmanager.rules.yaml:50]: is there a reason for dig instead

```suggestion
        severity: {{ .Values.customRules.AlertmanagerFailedReload.severity | default "critical" }}
```

I would prefer `severity` instead `labelsSeverity`

this pattern is more commonly used.

**Review by jkroepke — CHANGES_REQUESTED** [review]
Hi,

thanks for you contribution. I personally like the idea and I could also benefit from it.

However all files insides rules-1.14 directory are auto-generated by an [python script](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/hack/sync_prometheus_rules.py). In case I would merge this, the changes are getting reverted by next person who execution the script.

Please adjust the script https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/hack/sync_prometheus_rules.py to include your changes.

File: charts/kube-prometheus-stack/templates/prometheus/rules-1.14/alertmanager.rules.yaml (hunk 1 above)
Comment (ps-xaf) [reply charts/kube-prometheus-stack/templates/prometheus/rules-1.14/alertmanager.rules.yaml:50]: I tried different approaches but preferred dig, as it does not require to set the keys in the values yaml. So we don't end up with an entry for each alert rule like
```
customRules:
  AlertmanagerFailedReload: {}
  AlertmanagerMembersInconsistent: {}
  ...
```
With not set keys a nil pointer error is thrown.
```
Error: template: kube-prometheus-stack/templates/prometheus/rules-1.14/alertmanager.rules.yaml:45:21: executing "kube-prometheus-stack/templates/prometheus/rules-1.14/alertmanager.rules.yaml" at <.V
alues.customRules.AlertmanagerFailedReload.for>: nil pointer evaluating interface {}.for
```

Means each level of the keys had to be checked if it exists, which makes the template harder to read / maintain.
"dig" solves that for you :)

I am happy to change it from labelsSeverity to severity. My reason for labelsSeverity was the helm approach on [flat/nested values](https://helm.sh/docs/chart_best_practices/values/#flat-or-nested-values) as it shows the levels a bit nicer like lables.Severity in the manifests, and you can derive the hierachy. But if you still prefer serverity I change it.

**Comment by ps-xaf** [comment]
> However all files insides rules-1.14 directory are auto-generated by an [python script](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/hack/sync_prometheus_rules.py). In case I would merge this, the changes are getting reverted by next person who execution the script.
> 
> Please adjust the script https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/hack/sync_prometheus_rules.py to include your changes.

Thanks for the swift review and reply and pointing me into the right direction. I'll have a look the coming days how I can add it to the sync script.

File: charts/kube-prometheus-stack/templates/prometheus/rules-1.14/alertmanager.rules.yaml (hunk 1 above)
Comment (ps-xaf) [reply charts/kube-prometheus-stack/templates/prometheus/rules-1.14/alertmanager.rules.yaml]: @jkroepke i've update the PR and used 'severity' instead of 'labelsSeverity' , as it keeps the sync script a bit easier.

**Comment by jkroepke** [comment]
Please solve the merge conflicts, otherwise, CI would not run

**Comment by jkroepke** [comment]
@GMartinez-Sisti Since this is a larger PR and related to Python, I also would like to hear your opinion. In general, I like the idea, but it also add a lot of complexity.

File: charts/kube-prometheus-stack/Chart.yaml [hunk 2]
```diff
@@ -23,7 +23,7 @@ name: kube-prometheus-stack
 sources:
   - https://github.com/prometheus-community/helm-charts
   - https://github.com/prometheus-operator/kube-prometheus
-version: 56.7.0
+version: 56.7.1
```
Comment (GMartinez-Sisti) [inline charts/kube-prometheus-stack/Chart.yaml:26]: This is more a feature than a fix, so we should bump minor:
```suggestion
version: 56.8.0
```

File: charts/kube-prometheus-stack/templates/prometheus/rules-1.14/kubernetes-storage.yaml [hunk 3]
```diff
@@ -35,7 +35,7 @@ spec:
 {{- if .Values.defaultRules.additionalRuleGroupAnnotations.kubernetesStorage }}
 {{ toYaml .Values.defaultRules.additionalRuleGroupAnnotations.kubernetesStorage | indent 8 }}
 {{- end }}
-        description: The PersistentVolume claimed by {{`{{`}} $labels.persistentvolumeclaim {{`}}`}} in Namespace {{`{{`}} $labels.namespace {{`}}`}} on Cluster {{`{{`}} $labels.cluster {{`}}`}} is only {{`{{`}} $value | humanizePercentage {{`}}`}} free.
+        description: The PersistentVolume claimed by {{`{{`}} $labels.persistentvolumeclaim {{`}}`}} in Namespace {{`{{`}} $labels.namespace {{`}}`}} {{`{{`}} with $labels.cluster -{{`}}`}} on Cluster {{`{{`}} . {{`}}`}} {{`{{`}}- end {{`}}`}} is only {{`{{`}} $value | humanizePercentage {{`}}`}} free.
```
Comment (GMartinez-Sisti) [inline charts/kube-prometheus-stack/templates/prometheus/rules-1.14/kubernetes-storage.yaml:38]: Is this change intentional? Looks like it's reverting changes applied on https://github.com/prometheus-community/helm-charts/pull/4151 (same for similar changes below)

The result looks like this:
```diff
<         description: The PersistentVolume claimed by {{ $labels.persistentvolumeclaim }} in Namespace {{ $labels.namespace }} on Cluster {{ $labels.cluster }} is only {{ $value | humanizePercentage }} free.
---
>         description: The PersistentVolume claimed by {{ $labels.persistentvolumeclaim }} in Namespace {{ $labels.namespace }} {{ with $labels.cluster -}} on Cluster {{ . }} {{- end }} is only {{ $value | humanizePercentage }} free.
```

**Review by GMartinez-Sisti — COMMENTED** [review]
Overall LGTM, I think being able to customise these values is a great feature 💪 . Please check the comments.

File: charts/kube-prometheus-stack/templates/prometheus/rules-1.14/prometheus.yaml [hunk 4]
```diff
@@ -256,20 +256,20 @@ spec:
         summary: Prometheus is not ingesting samples.
       expr: |-
         (
-          rate(prometheus_tsdb_head_samples_appended_total{job="{{ $prometheusJob }}",namespace="{{ $namespace }}"}[5m]) <= 0
```
Comment (ps-xaf) [inline charts/kube-prometheus-stack/templates/prometheus/rules-1.14/prometheus.yaml:259]: Not intendet.

File: charts/kube-prometheus-stack/templates/prometheus/rules-1.14/kubernetes-storage.yaml (hunk 3 above)
Comment (ps-xaf) [reply charts/kube-prometheus-stack/templates/prometheus/rules-1.14/kubernetes-storage.yaml:38]: Thanks @GMartinez-Sisti, this was unintentional. I'll need to have a look as I get those changes also on main, when running sync_prometheus_rules.py.

A further unintended change is in [prometheus.yaml](https://github.com/prometheus-community/helm-charts/pull/4225/files#r1493836396)

File: charts/kube-prometheus-stack/Chart.yaml (hunk 2 above)
Comment (ps-xaf) [reply charts/kube-prometheus-stack/Chart.yaml]: bumped up to 56.9.0

File: charts/kube-prometheus-stack/templates/prometheus/rules-1.14/kubernetes-storage.yaml (hunk 3 above)
Comment (ps-xaf) [reply charts/kube-prometheus-stack/templates/prometheus/rules-1.14/kubernetes-storage.yaml]: I've removed those changes. They are from upstream due to [PR 2339](https://github.com/prometheus-operator/kube-prometheus/pull/2339) as the sync script fetches from main

See also upstreamd diff [1](https://github.com/prometheus-operator/kube-prometheus/pull/2339/files#diff-dc8532ace2a75bfcc266c6d1095a60dd1e2ab848af85816057b52d538614bb16R365) [2](https://github.com/prometheus-operator/kube-prometheus/pull/2339/files#diff-dc8532ace2a75bfcc266c6d1095a60dd1e2ab848af85816057b52d538614bb16R385) [3](https://github.com/prometheus-operator/kube-prometheus/pull/2339/files#diff-dc8532ace2a75bfcc266c6d1095a60dd1e2ab848af85816057b52d538614bb16R407) [4](https://github.com/prometheus-operator/kube-prometheus/pull/2339/files#diff-dc8532ace2a75bfcc266c6d1095a60dd1e2ab848af85816057b52d538614bb16R427) [5](https://github.com/prometheus-operator/kube-prometheus/pull/2339/files#diff-dc8532ace2a75bfcc266c6d1095a60dd1e2ab848af85816057b52d538614bb16R449)

File: charts/kube-prometheus-stack/templates/prometheus/rules-1.14/prometheus.yaml (hunk 4 above)
Comment (ps-xaf) [reply charts/kube-prometheus-stack/templates/prometheus/rules-1.14/prometheus.yaml]: Same, it's from upstream [pr 2344](https://github.com/prometheus-operator/kube-prometheus/pull/2344) and [rule diff](https://github.com/prometheus-operator/kube-prometheus/pull/2344/files#diff-2a066447e0230293dd113775548de681d94adcaeb1b32f8f9ab465411c8b6770R111)

**Comment by ps-xaf** [comment]
> Is this an manual revert? [`bf74da9` (#4225)](https://github.com/prometheus-community/helm-charts/pull/4225/commits/bf74da906cad6891a9e9188e626c1def79aa1aee)

yep, this was manually done.

**Comment by jkroepke** [comment]
> > Is this an manual revert? [`bf74da9` (#4225)](https://github.com/prometheus-community/helm-charts/pull/4225/commits/bf74da906cad6891a9e9188e626c1def79aa1aee)
> 
> yep, this was manually done.

Huch, the CI should normally fail here.

https://github.com/prometheus-community/helm-charts/blob/412f4d3b6b145f2d5f6a3cf408c34141058f35e6/charts/kube-prometheus-stack/ci/lint.sh#L25

**Review by jkroepke — APPROVED** [review]
> Unrelated to the check, an actual run of the script sync_prometheus_rules.py results in more changes than in this pr. Because it fetches upstream files [e.g.manifests/prometheus-prometheusRule.yaml](https://github.com/prometheus-operator/kube-prometheus/pull/2344/files#diff-2a066447e0230293dd113775548de681d94adcaeb1b32f8f9ab465411c8b6770R111) from prometheus-operator/kube-prometheus, which results in the changes related to description and expressions.

The easiest thing here would be a separate PR which include only changes from upstream. Then, rebase the PR here.

We can also leave it as-is. I'm fine with both here.

Looked through the generated diff here, and it looks good to me. I have no real opinion about the new python code here. Its an script which lives in a folder called `hack`.

If @GMartinez-Sisti agree here, too. I guess we can merge this.

**Comment by ps-xaf** [comment]
> > Unrelated to the check, an actual run of the script sync_prometheus_rules.py results in more changes than in this pr. Because it fetches upstream files [e.g.manifests/prometheus-prometheusRule.yaml](https://github.com/prometheus-operator/kube-prometheus/pull/2344/files#diff-2a066447e0230293dd113775548de681d94adcaeb1b32f8f9ab465411c8b6770R111) from prometheus-operator/kube-prometheus, which results in the changes related to description and expressions.
> 
> 
> 
> The easiest thing here would be a separate PR which include only changes from upstream. Then, rebase the PR here.

This would be cleaner but I leave it with you. 

Regards my comment related to the lint step. Forget about it, the check looks fine. If I have some time I test it in my repo.

**Review by jkroepke — APPROVED** [review]
Thanks for the PR! I guess its an huge improvement for a lot of users.

**Comment by degisftw** [comment]
This is amazing! And just in time for when I needed to do just this :) Thanks @ps-xaf !
